pkg/controller/certificaterequests/sync.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2019 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificaterequests
+
+import (
+	"context"
+	"encoding/json"
+	"reflect"
+
+	"github.com/kr/pretty"
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/validation"
+	"github.com/jetstack/cert-manager/pkg/issuer"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	errorCertificateNotExists = "CertNotExists"
+	errorCertificateParse     = "CertParseError"
+
+	errorIssuerNotFound = "IssuerNotFound"
+	errorIssuerInit     = "IssuerInitError"
+
+	successCertificateIssued = "CertIssued"
+)
+
+func (c *Controller) Sync(ctx context.Context, cr *v1alpha1.CertificateRequest) (err error) {
+	c.metrics.IncrementSyncCallCount(ControllerName)
+
+	log := logf.FromContext(ctx)
+	dbg := log.V(logf.DebugLevel)
+
+	crCopy := cr.DeepCopy()
+	defer func() {
+		if _, saveErr := c.updateCertificateRequestStatus(ctx, cr, crCopy); saveErr != nil {
+			err = utilerrors.NewAggregate([]error{saveErr, err})
+		}
+	}()
+
+	dbg.Info("Fetching existing certificate signing request and certificate from certificate request",
+		"name", crCopy.ObjectMeta.Name)
+
+	issuerObj, err := c.helper.GetGenericIssuer(crCopy.Spec.IssuerRef, crCopy.Namespace)
+	if k8sErrors.IsNotFound(err) {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotFound, err.Error())
+		c.log.Error(err,
+			"issuer-name", crCopy.Spec.IssuerRef.Name,
+			"issuer-kind", crCopy.Spec.IssuerRef.Kind)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
```

Comment:
After we have obtained a reference to a GenericIssuer from the helper, we can then call `log = logf.WithRelatedResource(log, issuerObj)` to ensure that all log key fields are set properly from now on